plaidml2/exec/exec.h

```diff
@@ -108,65 +111,46 @@ class Binder {
   }
 
   Buffer input(const edsl::Tensor& tensor) {  //
-    return inputs_.at(edsl::TensorRef{tensor});
+    return inputs_.at(tensor);
   }
 
   Buffer output(const edsl::Tensor& tensor) {  //
-    return outputs_.at(edsl::TensorRef{tensor});
+    return outputs_.at(tensor);
   }
 
   Binder& set_input(const edsl::Tensor& tensor, const Buffer& buffer) {
-    inputs_[edsl::TensorRef{tensor}] = buffer;
-    return *this;
-  }
-
-  Binder& set_inputs(const std::vector<edsl::Tensor>& inputs) {
-    for (auto input : inputs) {
-      set_input(input);
-    }
+    inputs_[tensor] = buffer;
     return *this;
   }
 
   Binder& set_output(const edsl::Tensor& tensor, const Buffer& buffer) {
-    outputs_[edsl::TensorRef{tensor}] = buffer;
-    return *this;
-  }
-
-  Binder& set_outputs(const std::vector<edsl::Tensor>& outputs) {
-    for (auto output : outputs) {
-      set_output(output);
-    }
+    outputs_[tensor] = buffer;
     return *this;
   }
 
-  Binder& set_input(const edsl::Tensor& tensor) {
-    auto shape = tensor.shape();
-    TensorShape tensor_shape(shape.dtype(), shape.int_dims());
-    return set_input(tensor, Buffer{device_, tensor_shape});
-  }
-
-  Binder& set_output(const edsl::Tensor& tensor) {
-    auto shape = tensor.shape();
-    TensorShape tensor_shape(shape.dtype(), shape.int_dims());
-    return set_output(tensor, Buffer{device_, tensor_shape});
-  }
-
   std::shared_ptr<Executable> compile() {
     std::vector<Binding> input_bindings;
-    for (const auto& kvp : inputs_) {
-      input_bindings.emplace_back(Binding{kvp.first.tensor, kvp.second});
-    }
     std::vector<Binding> output_bindings;
-    for (auto output : program_.outputs()) {
-      edsl::TensorRef ref{output};
-      auto it = outputs_.find(ref);
-      if (it == outputs_.end()) {
-        set_output(output);
-        it = outputs_.find(ref);
+    for (const auto& arg : program_.args()) {
+      if (arg.is_input) {
+        input_bindings.emplace_back(Binding{arg.tensor, get_buffer(&inputs_, arg)});
+      } else {
+        output_bindings.emplace_back(Binding{arg.tensor, get_buffer(&outputs_, arg)});
       }
-      output_bindings.emplace_back(Binding{output, it->second});
     }
-    return std::make_shared<Executable>(program_, input_bindings, output_bindings);
+    return std::make_shared<Executable>(program_, device_, target_, input_bindings, output_bindings);
+  }
+
+ private:
+  Buffer get_buffer(BindingMap* map, const edsl::ProgramArgument& arg) {
+    auto it = map->find(arg.tensor);
+    if (it != map->end()) {
+      return it->second;
+    }
+    TensorShape shape(arg.shape.dtype(), arg.shape.int_dims());
+    Buffer buffer{device_, shape};
+    map->emplace(arg.tensor, buffer);
```

Comment:
Eh, I'm returning the `buffer` anyways, and it's a `BufferPtr`, so it's cheap to make copies. Also, the `shape` wouldn't benefit from a `std::move` because the `Buffer` takes a `const TensorShape&`.

Reply:
Rather, the `Buffer` contains a `shared_ptr<plaidml_buffer>` so cheap to copy.